GVRf/Framework/src/org/gearvrf/GVRViewManager.java

```diff
@@ -320,6 +326,128 @@ public void run() {
+            centerCameraObject.attachCamera(centerCamera);
+            mainCameraRig.getOwnerObject().addChildObject(centerCameraObject);
+            GVRTransform centerCameraTransform = centerCameraObject
+                    .getTransform();
+
+            int index = 0;
+            // render +x face (the one user is currently looking at)
+            centerCameraTransform.rotateByAxis(-90, 0, 1, 0);
+            renderOneCameraAndAddToList(centerCamera, readbackBuffer,
+                    byteArrayArray, index++);
+
+            // render -x face
+            centerCameraTransform.rotateByAxis(180, 0, 1, 0);
+            renderOneCameraAndAddToList(centerCamera, readbackBuffer,
+                    byteArrayArray, index++);
+
+            // render +y face
+            centerCameraTransform.rotateByAxis(-90, 0, 1, 0);
+            centerCameraTransform.rotateByAxis(90, 1, 0, 0);
+            renderOneCameraAndAddToList(centerCamera, readbackBuffer,
+                    byteArrayArray, index++);
+
+            // render -y face
+            centerCameraTransform.rotateByAxis(180, 1, 0, 0);
+            renderOneCameraAndAddToList(centerCamera, readbackBuffer,
+                    byteArrayArray, index++);
+
+            // render +z face
+            centerCameraTransform.rotateByAxis(90, 1, 0, 0);
+            centerCameraTransform.rotateByAxis(180, 0, 1, 0);
+            renderOneCameraAndAddToList(centerCamera, readbackBuffer,
+                    byteArrayArray, index++);
+
+            // render -z face
+            centerCameraTransform.rotateByAxis(180, 0, 1, 0);
+            renderOneCameraAndAddToList(centerCamera, readbackBuffer,
+                    byteArrayArray, index++);
+
+            centerCameraObject.detachCamera();
+            mainCameraRig.getOwnerObject()
+                    .removeChildObject(centerCameraObject);
+        }
+    }
+
+    Object testObject = new Object();
+
+    void returnScreenshot3DToCaller(final GVRScreenshot3DCallback callback,
+            final byte[][] byteArrayArray, final int width, final int height) {
+
+        if (byteArrayArray.length != 6) {
+            throw new IllegalArgumentException(
+                    "byteArrayArray length is not 6.");
+        } else {
+            // run the callback function in a background thread
+            Threads.spawn(new Runnable() {
+                public void run() {
+                    final Bitmap[] bitmapArray = new Bitmap[6];
+                    Runnable[] threads = new Runnable[6];
+
+                    for (int i = 0; i < 6; i++) {
+                        final int index = i;
+                        threads[i] = new Runnable() {
+                            public void run() {
+                                byte[] bytearray = byteArrayArray[index];
+                                byteArrayArray[index] = null;
+                                bitmapArray[index] = generateBitmap(bytearray,
```

Comment:
Almost!

Imagine line 431 sees `bitmapArray[i] == null` *just* before line 416 sets it - we might still do the `wait()` after the `notify()`. This stuff can be really tricky, but I'm 99% sure that setting a local on line 416 and setting bitmapArray[index] *within* the synchronized block will avoid any race conditions.

    Bitmap face = generateBitmap(bytearray, width, height);
    synchronized (this) {
      bitmapArray[index] - face;
      notify();
    }